src/kernels/MIOpenLayerNorm.cpp

```diff
@@ -0,0 +1,259 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifdef MIOPEN_BETA_API
+
+#define GET(buf, idx) buf[idx]
+#define SET(buf, idx, val) buf[idx] = val
+
+#define GET_VAL(x, idx) GET(x, idx)
+#define SET_VAL(x, idx, val) SET(x, idx, val)
+
+#if MIOPEN_USE_FP64 == 1
+#define FSTYPE double
+#define MIOPEN_TYPE double
+#elif MIOPEN_USE_FP32 == 1
+#define FSTYPE float
+#define MIOPEN_TYPE float
+#elif MIOPEN_USE_FP16 == 1
+#define FSTYPE half
+#define MIOPEN_TYPE half
+#elif MIOPEN_USE_BF16 == 1
+#define FSTYPE hip_bfloat16
+#define MIOPEN_TYPE hip_bfloat16
+#endif
+
+#define LOCAL_SIZE 256
+
+extern "C" __global__ void LayernormFwdContiguous(const MIOPEN_TYPE* __restrict__ x,
+                                                  MIOPEN_TYPE* __restrict__ y,
+                                                  const MIOPEN_TYPE* __restrict__ weight,
+                                                  const MIOPEN_TYPE* __restrict__ bias,
+                                                  MIOPEN_TYPE* __restrict__ mean,
+                                                  MIOPEN_TYPE* __restrict__ rstd,
+                                                  float eps,
+                                                  uint64_t inner_size,
+                                                  bool mode)
+{
+    /*
+     * Each group works on a single channel.
+     * Example)
+     * x dim = {N, C, L}, normalized shape = {C, L}
+     * outer_size = N, inner_size = C * L
+     *
+     * Example2)
+     * x dim = {N, C, L}, normalized shape = {L}
+     * outer_size = N * C, inner_size = L
+     *
+     * => gws = {outer_size * LOCAL_SIZE}, lws = {LOCAL_SIZE}
+     */
+
+    /*
+     * Reduction to calculate mean and rstd
+     */
+
+    const uint64_t gid = blockIdx.x;
+    const uint64_t lid = threadIdx.x;
+
+    FSTYPE pmean = MIOPEN_TYPE(0);
```

Comment:
Here we have potentially different types and thus having two different defines for the same entity can be misleading.